src/azul/service/responseobjects/storage_service.py

```diff
@@ -39,12 +43,192 @@ def put(self, object_key: str, data: bytes, content_type: Optional[str] = None)
+
+        last_part = self.parts[-1]
+
+        if not last_part.already_uploaded:
+            # Per documentation, the last part can be at any size. Hence, the uploadable condition is ignored.
+            self.futures.append(self.thread_pool.submit(self._upload_part, last_part))
+
+        for _ in as_completed(self.futures):
+            pass  # Blocked until all uploads are done.
+
+        self.mp_upload.complete(MultipartUpload={"Parts": [part.to_dict() for part in self.parts]})
+        self.mp_upload = None
+        self.closed = True
+
+    def abort(self, raise_exception:bool=True):
+        """
+        Aborts a multipart upload.
+
+        :param raise_exception: Flag to raise exception on successful abort.
+        :raises UnexpectedMultipartUploadAbort: An abort due to unexpected
+                                                error occurred
+
+        If ``raise_exception`` is ``True``, :class:`UnexpectedMultipartUploadAbort`
+        will be raised.
+        """
+        if not self.is_active:
+            return
+
+        self.mp_upload.abort()
+        self.mp_upload = None
+        self.closed = True
+
+        logger.warning('Upload %s: Aborted', self.upload_id)
+        if raise_exception:
+            raise UnexpectedMultipartUploadAbort(f'{self.bucket_name}/{self.object_key}')
+
+    def push(self, data: bytes):
+        part = self._get_next_part(data)
+
+        if not part.is_uploadable:
+            return
+
+        part.uploaded = True
+
+        self.futures.append(self.thread_pool.submit(self._upload_part, part))
+
+    def _create_new_part(self, data: bytes):
+        part = Part(part_number=self.next_part_number, etag=None, content=[data], uploaded=False)
+        self.parts.append(part)
+        self.next_part_number += 1
+
+        return part
+
+    def _get_next_part(self, data):
+        if not self.parts:
+            part = self._create_new_part(data)
+        else:
+            # If the last part is under the minimum limit, the data will be appended.
+            last_part = self.parts[-1]
+            if last_part.already_uploaded:
+                part = self._create_new_part(data)
+            else:
+                part = last_part
+                part.content.append(data)
+
+        return part
+
+    def _upload_part(self, part):
+        upload_part = self.mp_upload.Part(part.part_number)
+        result = upload_part.upload(Body=b''.join(part.content))
+        part.etag = result['ETag']
+
+
+@dataclass
+class Part:
+    etag: str  # If ETag is defined, the content is already pushed to S3.
+    part_number: int
+    content: List[bytes]
+    uploaded: bool  # If true, the content is either being uploaded or already pushed to S3 (with etag defined).
+
+    @property
+    def already_uploaded(self):
+        return self.uploaded or self.etag is not None
+
+    @property
+    def is_uploadable(self):
+        return len(b''.join(self.content)) >= AWS_S3_DEFAULT_MINIMUM_PART_SIZE
```

Comment:
This is inefficient. If you only want the length, what you want to do is `sum(map(len,self.content)))`

Reply:
Then again I don't see the need for tracking incomplete parts in the first place.

Reply:
Originally, this code is to prevent `MultipartUploadHandler` from uploading less-than-5MB parts. Since we have ByteBuffer, there are two approaches to decide the fate of this property.
* Heavily rely on ByteBuffer to prevent that from happening
* Keep this as the final safeguard against unintentional upload.

Reply:
It's good to have a second line of defense against coding errors. But the defense shouldn't consist of duplicating logic or trying to gracefully handle the scenario you're trying to prevent. That would only exacerbate the problem by hiding it. The defense should be to raise an assertion error so we are made aware of the problem and can fix it. Besides, the assertion of an invariant is simpler and easier to implement than gracefully handling a violation of an invariant.